File: LogisimPlus/LogisimPlus/logisim/instance/Port.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.instance
{
	using EndData = logisim.comp.EndData;
	using logisim.data;
	using AttributeSet = logisim.data.AttributeSet;
	using BitWidth = logisim.data.BitWidth;
	using Location = logisim.data.Location;
	using StringGetter = logisim.util.StringGetter;

	public class Port
	{
		public const string INPUT = "input";
		public const string OUTPUT = "output";
		public const string INOUT = "inout";

		public const string EXCLUSIVE = "exclusive";
		public const string SHARED = "shared";

		private int dx;
		private int dy;
		private int type;
		private BitWidth widthFixed;
		private Attribute<BitWidth> widthAttr;
		private bool exclude;
		private StringGetter toolTip;

		public Port(int dx, int dy, string type, BitWidth bits) : this(dx, dy, type, bits, defaultExclusive(type))
		{
		}

		public Port(int dx, int dy, string type, int bits) : this(dx, dy, type, BitWidth.create(bits), defaultExclusive(type))
		{
		}

		public Port(int dx, int dy, string type, int bits, string exclude) : this(dx, dy, type, BitWidth.create(bits), exclude)
		{
		}

		public Port(int dx, int dy, string type, BitWidth bits, string exclude)
		{
			this.dx = dx;
			this.dy = dy;
			this.type = toType(type);
			this.widthFixed = bits;
			this.widthAttr = null;
			this.exclude = toExclusive(exclude);
			this.toolTip = null;
		}

		public Port(int dx, int dy, string type, Attribute<BitWidth> attr) : this(dx, dy, type, attr, defaultExclusive(type))
		{
		}

		public Port(int dx, int dy, string type, Attribute<BitWidth> attr, string exclude)
		{
			this.dx = dx;
			this.dy = dy;
			this.type = toType(type);
			this.widthFixed = null;
			this.widthAttr = attr;
			this.exclude = toExclusive(exclude);
			this.toolTip = null;
		}

		public virtual StringGetter ToolTip
		{
			set
			{
				toolTip = value;
			}
			get
			{
				StringGetter getter = toolTip;
				return getter == null ? null : getter.get();
			}
		}


		public virtual Attribute<BitWidth> WidthAttribute
		{
			get
			{
				return widthAttr;
			}
		}

		public virtual EndData toEnd(Location loc, AttributeSet attrs)
		{
			Location pt = loc.translate(dx, dy);
			if (widthFixed != null)
			{
				return new EndData(pt, widthFixed, type, exclude);
			}
			else
			{
				object val = attrs.getValue(widthAttr);
				if (!(val is BitWidth))
				{
					throw new System.ArgumentException("Width attribute not set");
				}
				return new EndData(pt, (BitWidth) val, type, exclude);
			}
		}

		private static int toType(string s)
		{
			if (string.ReferenceEquals(s, null))
			{
				throw new System.ArgumentException("Null port type");
			}
			else if (s.Equals(INPUT))
			{
				return EndData.INPUT_ONLY;
			}
			else if (s.Equals(OUTPUT))
			{
				return EndData.OUTPUT_ONLY;
			}
			else if (s.Equals(INOUT))
			{
				return EndData.INPUT_OUTPUT;
			}
			else
			{
				throw new System.ArgumentException("Not recognized port type");
			}
		}

		private static string defaultExclusive(string s)
		{
			if (string.ReferenceEquals(s, null))
			{
				throw new System.ArgumentException("Null port type");
			}
			else if (s.Equals(INPUT))
			{
				return SHARED;
			}
			else if (s.Equals(OUTPUT))
			{
				return EXCLUSIVE;
			}
			else if (s.Equals(INOUT))
			{
				return SHARED;
			}
			else
			{
				throw new System.ArgumentException("Not recognized port type");
			}
		}

		private static bool toExclusive(string s)
		{
			if (string.ReferenceEquals(s, null))
			{
				throw new System.ArgumentException("Null exclusion type");
			}
			else if (s.Equals(EXCLUSIVE))
			{
				return true;
			}
			else if (s.Equals(SHARED))
			{
				return false;
			}
			else
			{
				throw new System.ArgumentException("Not recognized exclusion type");
			}
		}
	}

}

File: logisim/app/src/main/java/logisim/instance/Port.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.instance;

import logisim.comp.EndData;
import logisim.data.Attribute;
import logisim.data.AttributeSet;
import logisim.data.BitWidth;
import logisim.data.Location;
import logisim.util.StringGetter;

public class Port {
	public static final String INPUT = "input";
	public static final String OUTPUT = "output";
	public static final String INOUT = "inout";

	public static final String EXCLUSIVE = "exclusive";
	public static final String SHARED = "shared";

	private int dx;
	private int dy;
	private int type;
	private BitWidth widthFixed;
	private Attribute<BitWidth> widthAttr;
	private boolean exclude;
	private StringGetter toolTip;

	public Port(int dx, int dy, String type, BitWidth bits) {
		this(dx, dy, type, bits, defaultExclusive(type));
	}

	public Port(int dx, int dy, String type, int bits) {
		this(dx, dy, type, BitWidth.create(bits), defaultExclusive(type));
	}

	public Port(int dx, int dy, String type, int bits, String exclude) {
		this(dx, dy, type, BitWidth.create(bits), exclude);
	}

	public Port(int dx, int dy, String type, BitWidth bits, String exclude) {
		this.dx = dx;
		this.dy = dy;
		this.type = toType(type);
		widthFixed = bits;
		widthAttr = null;
		this.exclude = toExclusive(exclude);
		toolTip = null;
	}

	public Port(int dx, int dy, String type, Attribute<BitWidth> attr) {
		this(dx, dy, type, attr, defaultExclusive(type));
	}

	public Port(int dx, int dy, String type, Attribute<BitWidth> attr, String exclude) {
		this.dx = dx;
		this.dy = dy;
		this.type = toType(type);
		widthFixed = null;
		widthAttr = attr;
		this.exclude = toExclusive(exclude);
		toolTip = null;
	}

	public void setToolTip(StringGetter value) {
		toolTip = value;
	}

	public String getToolTip() {
		StringGetter getter = toolTip;
		return getter == null ? null : getter.get();
	}

	public Attribute<BitWidth> getWidthAttribute() {
		return widthAttr;
	}

	public EndData toEnd(Location loc, AttributeSet attrs) {
		Location pt = loc.add(dx, dy);
		if (widthFixed != null)
			return new EndData(pt, widthFixed, type, exclude);
		BitWidth val = attrs.getValue(widthAttr);
		if (val == null)
			throw new IllegalArgumentException("Width attribute not set");
		return new EndData(pt, val, type, exclude);
	}

	private static int toType(String s) {
		if (s == null)
			throw new IllegalArgumentException("Null port type");
		else if (s.equals(INPUT))
			return EndData.INPUT_ONLY;
		else if (s.equals(OUTPUT))
			return EndData.OUTPUT_ONLY;
		else if (s.equals(INOUT))
			return EndData.INPUT_OUTPUT;
		else
			throw new IllegalArgumentException("Not recognized port type");
	}

	private static String defaultExclusive(String s) {
		if (s == null)
			throw new IllegalArgumentException("Null port type");
		else if (s.equals(INPUT))
			return SHARED;
		else if (s.equals(OUTPUT))
			return EXCLUSIVE;
		else if (s.equals(INOUT))
			return SHARED;
		else
			throw new IllegalArgumentException("Not recognized port type");
	}

	private static boolean toExclusive(String s) {
		if (s == null)
			throw new IllegalArgumentException("Null exclusion type");
		else if (s.equals(EXCLUSIVE))
			return true;
		else if (s.equals(SHARED))
			return false;
		else
			throw new IllegalArgumentException("Not recognized exclusion type");
	}
}
